MCR 2997 fix pmd issues in mycore base

Antoniafriedrich opened the pull request:
[Link to jira](https://mycore.atlassian.net/browse/MCR-2997).

Commit:
MCR-2997 AvoidThrowingNullPointerException

Commit:
MCR-2997 MethodNamingConventions

Commit:
MCR-2997 BrokenNullCheck

Commit:
MCR-2997 AvoidReassigningParameters

Commit:
MCR-2997 CompareObjectsWithEquals

Commit:
MCR-2997 FieldDeclarationsShouldBeAtStartOfClass

Commit:
MCR-2997 AppendCharacterWithChar

Commit:
MCR-2997 CollapsibleIfStatements

Commit:
MCR-2997 SimplifiableTestAssertion

Commit:
MCR-2997 UnnecessaryBooleanAssertion

Commit:
MCR-2997 InefficientStringBuffering

Commit:
MCR-2997 UnusedFormalParameter

Commit:
MCR-2997 AvoidThrowingRawExceptionTypes

Commit:
MCR-2997 AvoidPrintStackTrace

Commit:
MCR-2997 AvoidCatchingThrowable

Commit:
MCR-2997 UnusedPrivateMethod

Commit:
MCR-2997 UnnecessaryReturn

Commit:
MCR-2997 SimplifyConditional

Commit:
MCR-2997 UnnecessaryConstructor

Commit:
MCR-2997 ImplicitSwitchFallThrough

Commit:
MCR-2997 UnusedLocalVariables

Commit:
MCR-2997 AvoidThrowingNewInstanceOfSameException

Commit:
MCR-2997 AssignmentToNoneFinalStatic

Commit:
MCR-2997 SingularField

Commit:
MCR-2997 UnnecessaryImport

Commit:
MCR-2997 OneDeclarationPerLine

Commit:
MCR-2997 UnnecessaryFullyQualifiedName

Commit:
MCR-2997 NPathComplexity

Commit:
MCR-2997 SimplifyBooleanReturns

Commit:
MCR-2997 UnusedPrivateField

Commit:
MCR-2997 UncommentedEmptyMethod

Commit:
MCR-2997 NcssCount

Commit:
Merge remote-tracking branch 'origin/main' into issues/MCR-2997-fix-pmd-issues-in-mycore-base

Commit:
fix Codacy problems

toKrause commented on mycore-base/src/main/java/org/mycore/datamodel/metadata/validator/MCREditorOutValidator.java:
```diff
@@ -202,7 +202,7 @@ public static String checkMetaObjectWithLangNotEmpty(Element datasubtag,
     public static String checkMetaObjectWithLinks(Element datasubtag, Class<? extends MCRMetaInterface> metaClass) {
         if (datasubtag.getAttributeValue("href") == null
-            && datasubtag.getAttributeValue("href", XLINK_NAMESPACE) == null) {
+            || datasubtag.getAttributeValue("href", XLINK_NAMESPACE) == null) {
```
I think the conditional "no `href` and no `xlink:href`" was actually correct. Otherwise the code in lines 215 - 218 would be useless.

toKrause commented on mycore-base/src/main/java/org/mycore/common/content/util/MCRServletContentHelper.java:
```diff
@@ -274,9 +274,7 @@ private static boolean checkIfMatch(final HttpServletRequest request, final Http
-        if (headerValue != null) {
-            if (headerValue.indexOf('*') == -1) {
-
+        if (headerValue != null&&headerValue.indexOf('*') == -1) {
```
Almost all changes due to `CollapsibleIfStatements`  need their formatting fixed (should have spaces around `&&`, I've added comments the the other occurrences).

toKrause commented on mycore-base/src/main/java/org/mycore/common/config/MCRInstanceConfiguration.java:
```diff
@@ -135,12 +135,10 @@ private static Map<String, String> reduceProperties(MCRInstanceName name, String
             String value = entry.getValue();
-            if (key.startsWith(prefix)) {
-                if (key.length() != prefix.length() && key.charAt(prefix.length()) == '.') {
+            if (key.startsWith(prefix)&&key.length() != prefix.length() && key.charAt(prefix.length()) == '.') {
```
The condition was intentionally split into two parts, to make it easier to follow the thought-process.

The latter two checks only make sense in the case, where the first condition is met.

toKrause commented on mycore-base/src/main/java/org/mycore/common/xml/MCRLayoutService.java:
```diff
@@ -117,7 +117,9 @@ public MCRContent getTransformedContent(HttpServletRequest req, HttpServletRespo
             String filename = getFileName(req, parameter);
-            return transform(transformer, source, parameter, filename);
+            MCRContent transformed = transform(transformer, source, parameter);
+            transformed.setName(filename);
```
Where does this line com from? Should this really be part of a refactoring to fix the `UnusedFormalParameter` issue?

This might be an actual bug that should be handled in a separate tocket.

toKrause commented on mycore-base/src/main/java/org/mycore/common/content/MCRStringContent.java:
```diff
@@ -44,7 +45,7 @@ public MCRStringContent(String text) {
             setEncoding(MCRConstants.DEFAULT_ENCODING);
         } catch (UnsupportedEncodingException e) {
-            throw new RuntimeException(e);
+            throw new IllegalArgumentException(e);
```
I'm not convinced that  an `IllegalArgumentException` is correct here. Whether or not the exception is thrown doesn't appear to depend on the functions argument.

toKrause commented on mycore-base/src/main/java/org/mycore/common/content/MCRXMLContent.java:
```diff
@@ -58,7 +58,7 @@ public MCRXMLContent() {
             this.setEncoding(MCRConstants.DEFAULT_ENCODING);
         } catch (UnsupportedEncodingException e) {
-            throw new RuntimeException(e);
+            throw new IllegalArgumentException(e);
```
I'm not convinced that  an `IllegalArgumentException` is correct here. Whether or not the exception is thrown doesn't appear to depend on the functions argument.

toKrause replied:
This one thing still needs attention .

It's the same detail that was already improved in `MCRStringContent`, i.e. should be `throw new MCRException("MCRXMLContent encoding not supported", e);`. Probably drowned out between all the other stuff and because GitHub, for whatever reason, decided to hide the conversation in the timeline.

Antoniafriedrich replied:
Strange. I think I corrected that already. Anyway, it should be now.

toKrause commented on mycore-base/src/main/java/org/mycore/common/xml/MCRXMLFunctions.java:
```diff
@@ -272,20 +271,6 @@ public static String regexp(String source, String regex, String replace) {
-     *
-     * @return the string matching in the source parameter matching the given regular expression
-     */
-    public static String getMatchingString(String source, String regex) {
```
Why was this method removed as part of fixing the `AvoidCatchingThrowable` issue?

toKrause commented on mycore-base/src/main/java/org/mycore/common/xml/MCRXMLFunctions.java:
```diff
@@ -880,20 +857,6 @@ public static String stripHtml(final String s) {
-     * @return approximately the usable space in the MCR.datadir
-     * @throws IOException
-     * */
-    public static long getUsableSpace() throws IOException{
```
Why was this method removed as part of fixing the `AvoidCatchingThrowable` issue?

toKrause commented on mycore-base/src/main/java/org/mycore/common/xml/MCRXMLFunctions.java:
```diff
@@ -530,16 +515,8 @@ public static String getMimeType(String f) {
             return "application/octet-stream";
         }
-
-        if (MIMETYPE_MAP == null) {
```
Why was the caching of the `MimetypesFileTypeMap`  removed as part of fixing the `AvoidCatchingThrowable` issue?

toKrause commented on mycore-base/src/main/java/org/mycore/datamodel/niofs/MCRFileSystemPromoter.java:
```diff
@@ -82,7 +82,8 @@ public void startUp(ServletContext servletContext) {
         }
     }
-
+@SuppressWarnings("PMD.UnusedPrivateMethod")
```
Incorrect indentation.

toKrause commented on mycore-base/src/main/java/org/mycore/datamodel/common/MCRDataURL.java:
```diff
@@ -273,7 +273,7 @@ public static MCRDataURL parse(final String dataURL) throws MalformedURLExceptio
                         try {
                             return decode(sl[1], StandardCharsets.UTF_8);
-                        } catch (UnsupportedCharsetException e) {
+                        } catch (Exception e) {
```
This change is a cheat that worsens the code in order to get rid of a PMD warning. The only possible exception still is an `UnsupportedCharsetException`, but now the code reads as if other exceptions might be possible.

toKrause commented on mycore-base/src/main/java/org/mycore/backend/jpa/links/MCRLINKHREFPK.java:
```diff
@@ -139,37 +140,23 @@ public boolean equals(Object obj) {
     }
 
+    private boolean stringsAreEqual(String string, String otherString) {
+        return (!((string == null && otherString != null) || !Objects.equals(string, otherString)));
```
`Objects#equals` should already handle the `null`-checks. No need for the first

toKrause commented on mycore-base/src/main/java/org/mycore/backend/jpa/access/MCRACCESSRULE.java:
```diff
@@ -103,52 +104,25 @@ public boolean equals(Object obj) {
+    }
+
+    private boolean stringsAreEqual(String string, String otherString) {
+        return (!((string == null && otherString != null) || !Objects.equals(string, otherString)));
```
`Objects#equals` should already handle the `null`-checks. No need for the first

toKrause commented on mycore-base/src/main/java/org/mycore/common/xml/MCRURIResolver.java:
```diff
@@ -114,6 +114,8 @@
 
 import jakarta.servlet.ServletContext;
 
+import static org.mycore.common.xml.MCRURIResolver.MCRLayoutTransformerResolver.findAndThrowTransformerException;
```
Statically importing a package private static method of a package private static subclass of the importing class in order to use in a different package private static subclass of the same class seems really weird.

Why not move the method to `MCRURIResolver`, so it can be used by all subclasses?

Commit:
Implements suggested changes and fixed formatting issues with the changes of "CollapsibleIfStatements"

Commit:
fixed some general issues

Commit:
Fix MCRACCESSRULE stringsAreEqual check

toKrause commented on mycore-base/src/main/java/org/mycore/common/config/MCRInstanceConfiguration.java:
```diff
@@ -129,16 +129,20 @@ public static MCRInstanceConfiguration ofName(MCRInstanceName name, Map<String,
     }
 
+    @SuppressWarnings("PMD.CollapsibleIfStatements")
+    //The condition was intentionally split into to warts, to make it to follow the thought-process.
```
My original comment was indeed garbled up, should have been:

> The condition was intentionally split into two parts, to make it easier to follow the thought-process.
> The latter two checks only make sense in the case, where the first condition is met.

Antoniafriedrich replied:
The comment was corrected in the following commit.

toKrause replied:
As far as I can tell, the comment still contains the unintelligible part "into to warts, to make it to follow the thought-process"

Commit:
Corrects a comment

toKrause commented on mycore-base/src/main/java/org/mycore/datamodel/common/MCRDataURL.java:
```diff
@@ -170,15 +173,12 @@ public static String build(final NodeList nodeList, final String encoding, final
-                } else {
-                    enc = MCRDataURLEncoding.BASE64;
-                }
+        if (mtm.matches()&&enc == null) {
```
Still needs formatting fixed.

toKrause commented on mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRMetaLink.java:
```diff
@@ -234,10 +234,8 @@ public String getXLinkRole() {
-            if (linktype.equals(input.getXLinkType()) && href.equals(input.getXLinkHref())) {
-                return true;
-            }
+        if (linktype.equals("locator")&&linktype.equals(input.getXLinkType()) && href.equals(input.getXLinkHref())) {
```
Still needs formatting fixed.

toKrause commented on mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRMetaLink.java:
```diff
@@ -404,10 +400,8 @@ public JsonObject createJSON() {
-            if (!XMLChar.isValidNCName(label)) {
-                throw new MCRException(getSubTag() + ": label is no valid NCName:" + label);
-            }
+        if (label != null && label.length() > 0&&!XMLChar.isValidNCName(label)) {
```
Still needs formatting fixed.

toKrause commented on mycore-base/src/main/java/org/mycore/datamodel/metadata/MCRMetaLink.java:
```diff
@@ -416,25 +410,22 @@ public void validate() throws MCRException {
-            if (href == null || href.length() == 0) {
-                throw new MCRException(getSubTag() + ": href is null or empty");
-            }
+        if (linktype.equals("locator")&&(href == null || href.length() == 0)) {
```
Still needs formatting fixed.

toKrause commented on mycore-base/src/main/java/org/mycore/frontend/cli/MCRBatchEditorCommands.java:
```diff
@@ -133,10 +133,8 @@ public static void edit(String oid, String level, Action a, String field1, Strin
-                if (!find(base, field1, value1).isEmpty()) {
-                    find(base, field2, value2).forEach(Element::detach);
-                }
+            } else if (a == Action.REMOVE_IF&&!find(base, field1, value1).isEmpty()) {
```
Still needs formatting fixed.

toKrause commented on mycore-base/src/main/java/org/mycore/tools/MCRTopologicalSort.java:
```diff
@@ -283,12 +283,10 @@ public int[] doTopoSort() {
-                        leafs.add(to);
-                    }
+                // remove edge e from the graph
+                if (ts != null && ts.contains(node)&&removeEdge(node, to)) {
```
Still needs formatting fixed.

Commit:
Fixed formatting issues with &&

Commit:
Fix issues with rulesparser and remove ClassnamingRule

Methods that were falsely removed are restored, MCRRuleParser errors were fixed,comment was corrected. Error with MCRTransformerPipe.transform was removed by undoing changes.

Antoniafriedrich committed:
Merge branch 'main' into issues/MCR-2997-fix-pmd-issues-in-mycore-base

Commit:
Change IllegalArgumentException to MCRException in MCRXMLContent
